src/fhirtypes/mixins.ts

```diff
@@ -53,27 +65,40 @@ export class HasId {
    * If the FHIRId does not match, log an error.
    * If the id is a valid name, sanitize it to a valid id and log a warning
    *
+   * @param {FshStructure | FshCodeSystem | FshValueSet} fshDefinition - The entity who's id is being set
    * @param sourceInfo - The FSH file and location that specified the id
    */
-  validateId(sourceInfo: SourceInfo) {
+  validateId(fshDefinition: FshStructure | FshCodeSystem | FshValueSet) {
     let validId = idRegex.test(this.id);
-    if (!validId && nameRegex.test(this.id)) {
-      // A valid name can be turned into a valid id by replacing _ with - and slicing to 64 character limit
-      const sanitizedId = this.id.replace(/_/g, '-').slice(0, 64);
-      if (idRegex.test(sanitizedId)) {
-        // Use the sanitized id, but warn the user to fix this
-        logger.warn(
-          `The string "${this.id}" represents a valid FHIR name but not a valid FHIR id. FHIR ids cannot contain "_" and can be at most 64 characters. The id will be exported as "${sanitizedId}". Avoid this warning by specifying a valid id directly using the "Id" keyword.`,
-          sourceInfo
-        );
-        this.id = sanitizedId;
+    const idRule = fshDefinition.rules
+      .filter(rule => rule instanceof CaretValueRule && rule.caretPath === 'id')
+      .pop() as CaretValueRule;
+    if (!validId) {
+      if (idRule && idRegex.test(idRule.value as string)) {
```

Comment:
I think a similar simplification can probably happen here as I suggested for name.